sdk/nodejs/appservice/zMixins.ts

```diff
@@ -246,81 +255,257 @@ export interface HostSettings {
+}
+
+function makeSafeStorageAccountName(prefix: string) {
+    // Account name needs to be at max 24 chars (minus the extra 8 random chars);
+    // not exceed the max length of 24.
+    // Name must be alphanumeric.
+    return prefix.replace(/[^a-zA-Z0-9]/g, "").toLowerCase().substr(0, 24 - 8);
+}
+
+function makeSafeStorageContainerName(prefix: string) {
+    // Account name needs to be at max 63 chars (minus the extra 8 random chars);
+    // Name must be alphanumeric (and hyphens).
+    return prefix.replace(/[^a-zA-Z0-9-]/g, "").toLowerCase().substr(0, 63 - 8);
+}
+
+interface BaseSubscriptionArgs {
+    resourceGroup?: core.ResourceGroup;
+    resourceGroupName?: pulumi.Input<string>;
+    location?: pulumi.Input<string>;
+}
+
+/** @internal */
+export function getResourceGroupNameAndLocation(
+        args: BaseSubscriptionArgs, fallbackResourceGroupName: pulumi.Output<string> | undefined) {
+
+    if (args.resourceGroup) {
+        return { resourceGroupName: args.resourceGroup.name, location: args.resourceGroup.location };
+    }
+
+    if (!fallbackResourceGroupName) {
+        throw new Error("Either [args.resourceGroup] or [args.resourceGroupName] must be provided.");
+    }
+
+    const resourceGroupName = util.ifUndefined(args.resourceGroupName, fallbackResourceGroupName);
+    const getResult = resourceGroupName.apply(n => core.getResourceGroup({ name: n }));
+    return { resourceGroupName, location: getResult.location };
+}
+
+export type CallbackArgs<C extends Context<R>, E, R extends Result> = {
+    /**
+     * The Javascript function instance to use as the entrypoint for the Azure FunctionApp.  Either
+     * [callback] or [callbackFactory] must be provided.
+     */
+    callback?: Callback<C, E, R>;
+
+    /**
+     * The Javascript function instance that will be called to produce the function that is the
+     * entrypoint for the Azure FunctionApp. Either [callback] or [callbackFactory] must be
+     * provided.
+     *
+     * This form is useful when there is expensive initialization work that should only be executed
+     * once.  The factory-function will be invoked once when the final Azure FunctionApp module is
+     * loaded. It can run whatever code it needs, and will end by returning the actual function that
+     * the Azure will call into each time the FunctionApp it is is invoked.
+     */
+    callbackFactory?: CallbackFactory<C, E, R>;
+}
+
+/** @internal */
+export function serializeFunctionCallback<C extends Context<R>, E, R extends Result>(
+    args: CallbackArgs<C, E, R>): Promise<pulumi.runtime.SerializedFunction> {
 
     if (args.callback && args.callbackFactory) {
         throw new pulumi.RunError("Cannot provide both [callback] and [callbackFactory]");
     }
 
     if (!args.callback && !args.callbackFactory) {
-        throw new Error("One of [callback] or [callbackFactory] must be provided.");
+        throw new pulumi.RunError("One of [callback] or [callbackFactory] must be provided.");

```

Comment:
Note: run-error is wrong in both cases.  It should just be error.  RunError is basically when we do *not* want the stack, because we can just give a reasonable message that needs no additional context (like: you need to provide config value X).  For arg validation like this, we actually want a normal error so you get a good stack making it clear what codepath is causing you to hit this.